design/backup_cancellation.md

```diff
@@ -0,0 +1,154 @@
+
+# Backup Cancellation Design
+
+## Abstract
+This proposal introduces user-initiated backup cancellation functionality to Velero, allowing users to abort running backups through a new `cancel` field in the backup specification.
+The design addresses GitHub issues [#9189](https://github.com/vmware-tanzu/velero/issues/9189
+) and [#2098](https://github.com/vmware-tanzu/velero/issues/2098) by providing a mechanism to cleanly cancel async operations and prevent resource leaks when backups need to be terminated.
+
+## Background
+Currently, Velero lacks the ability to cancel running backups, leading to several critical issues.
+When users accidentally submit broad backup jobs (e.g., forgot to narrow resource selectors), the system becomes blocked and scheduled jobs accumulate.
+Additionally, the backup deletion controller prevents running backups from being deleted.
+
+
+## Goals
+- Enable users to cancel running backups through a `cancel` field in the backup specification
+- Cleanly cancel all associated async operations (BackupItemAction operations, DataUploads, PodVolumeBackups)
+- Provide clear backup phase transitions (InProgress → Cancelling → Cancelled)
+
+## Non Goals
+- Cancelling backups that have already completed or failed
+- Rolling back partially completed backup operations
+- Implementing cancellation for restore operations (future work)
+
+
+## High-Level Design
+The solution introduces a new `cancel` boolean field to the backup specification that users can set to `true` to request cancellation.
+Existing controllers (backup_controller, backup_operations_controller, backup_finalizer_controller) will check for this field and transition the backup to a `Cancelling` phase before returning early from their reconcile loops.
+
+A new dedicated backup cancellation controller will watch for backups in the `Cancelling` phase and coordinate the actual cancellation work.
```

Comment:
Why do we need a dedicated controller?
Logically, cancellation are phases of the backup CR as same as other phases, e.g., new, failed, inProgress, etc. It is more reasonable to process them all together.
Practically, if we allow multiple controllers to process the same backup CR, we must pay attention on the contest resolving, which is not necessary.

Reply:
We already have more than one controller for the backup. There's the backup reconciler and the backup operations reconciler. But I think you're right -- we don't necessarily need to create a completely new one. We may not even need both "Canceling" and "Cancelled" phases -- basically the existing controllers see "cancel=true", stop work, cancel datauploads, set phase="cancelled"

Reply:
The dedicated controller could be avoid if the following are handled within the same controllers:
1. async op cancellation: `backup_operations_controller` could do this 
2. backup data deletion: `backup_deletion_controller` would need to be invoked by creating a special `BackupDeleteRequest` that doesn't delete the logs and the backup resource itself

Would mean that multiple points in other controllers would need to do step 2 if cancel is true

Reply:
Cancel action and cancelled status is more suitable to be part of the primitive state machine for the backup CR, so the processing is better to be together with other primitive states, which is helpful for code understanding and code management.

Async operation is not with a action-state pattern and backup deletion controller is not for backup CR but for deleteBackupRequest CR.